File: scripts/enhanceImages.mjs
/**
 * 批量图片增强脚本
 * 使用豆包seedream4.5 AI接口将图片变得更清晰
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// API配置
const API_KEY = 'ark-723d46c8-07f1-4084-8750-5d45cb53eae5-c67eb';
const API_URL = 'https://ark.cn-beijing.volces.com/api/v3/images/generations';
const MODEL = 'doubao-seedream-4-5-251128';

// 图片目录配置
const EQUIPMENT_DIR = path.join(__dirname, '../public/images/equipment');
const OUTPUT_DIR = path.join(__dirname, '../public/images/equipment_enhanced');

// 支持的图片格式
const SUPPORTED_FORMATS = ['.png', '.jpg', '.jpeg', '.webp'];

// 延迟函数，避免API调用过快
const delay = (ms) => new Promise(resolve => setTimeout(resolve, ms));

/**
 * 将图片文件转换为base64格式
 * @param {string} imagePath - 图片文件路径
 * @returns {string} base64编码的图片数据
 */
function imageToBase64(imagePath) {
  const imageBuffer = fs.readFileSync(imagePath);
  const base64 = imageBuffer.toString('base64');
  const ext = path.extname(imagePath).toLowerCase();
  
  // 根据文件扩展名确定MIME类型
  const mimeTypes = {
    '.png': 'image/png',
    '.jpg': 'image/jpeg',
    '.jpeg': 'image/jpeg',
    '.webp': 'image/webp'
  };
  
  const mimeType = mimeTypes[ext] || 'image/png';
  return `data:${mimeType};base64,${base64}`;
}

/**
 * 调用豆包API增强图片
 * @param {string} imageBase64 - base64编码的图片
 * @param {string} imageName - 图片名称（用于生成prompt）
 * @returns {Promise<string>} 增强后的图片URL
 */
async function enhanceImage(imageBase64, imageName) {
  // 根据图片名称生成合适的prompt
  const prompt = `将这张游戏装备图片变得更清晰、更精致，保持原有风格和细节，提升画质和清晰度：${imageName}`;
  
  try {
    const response = await fetch(API_URL, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${API_KEY}`
      },
      body: JSON.stringify({
        model: MODEL,
        prompt: prompt,
        image: imageBase64,
        sequential_image_generation: 'disabled',
        response_format: 'url',
        size: '2K',
        stream: false,
        watermark: false
      })
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`API请求失败: ${response.status} - ${errorText}`);
    }

    const result = await response.json();
    
    // 检查返回的数据结构
    if (result.data && result.data[0] && result.data[0].url) {
      return result.data[0].url;
    } else if (result.url) {
      return result.url;
    } else {
      throw new Error(`API返回数据格式异常: ${JSON.stringify(result)}`);
    }
  } catch (error) {
    console.error(`增强图片失败 (${imageName}):`, error.message);
    throw error;
  }
}

/**
 * 下载图片并保存到本地
 * @param {string} url - 图片URL
 * @param {string} outputPath - 输出文件路径
 */
async function downloadImage(url, outputPath) {
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`下载图片失败: ${response.status}`);
  }
  
  const arrayBuffer = await response.arrayBuffer();
  const buffer = Buffer.from(arrayBuffer);
  
  // 确保输出目录存在
  const outputDir = path.dirname(outputPath);
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }
  
  fs.writeFileSync(outputPath, buffer);
}

/**
 * 获取所有需要处理的图片文件
 * @returns {Array<{category: string, file: string, path: string}>}
 */
function getAllImages() {
  const images = [];
  
  // 读取equipment目录下的所有子目录
  const categories = fs.readdirSync(EQUIPMENT_DIR, { withFileTypes: true })
    .filter(dirent => dirent.isDirectory())
    .map(dirent => dirent.name);
  
  // 遍历每个分类目录
  for (const category of categories) {
    const categoryPath = path.join(EQUIPMENT_DIR, category);
    const files = fs.readdirSync(categoryPath);
    
    for (const file of files) {
      const ext = path.extname(file).toLowerCase();
      if (SUPPORTED_FORMATS.includes(ext)) {
        images.push({
          category,
          file,
          path: path.join(categoryPath, file)
        });
      }
    }
  }
  
  return images;
}

/**
 * 主函数：批量处理所有图片
 */
async function main() {
  console.log('========================================');
  console.log('图片批量增强脚本启动');
  console.log('========================================\n');
  
  // 获取所有图片
  const images = getAllImages();
  console.log(`找到 ${images.length} 张图片需要处理\n`);
  
  if (images.length === 0) {
    console.log('没有找到需要处理的图片');
    return;
  }
  
  // 创建输出目录
  if (!fs.existsSync(OUTPUT_DIR)) {
    fs.mkdirSync(OUTPUT_DIR, { recursive: true });
  }
  
  // 统计信息
  let successCount = 0;
  let failCount = 0;
  const failedImages = [];
  
  // 处理每张图片
  for (let i = 0; i < images.length; i++) {
    const { category, file, path: imagePath } = images[i];
    const imageBase64 = imageToBase64(imagePath);
    
    console.log(`[${i + 1}/${images.length}] 正在处理: ${category}/${file}`);
    
    try {
      // 调用API增强图片
      const enhancedUrl = await enhanceImage(imageBase64, `${category}/${file}`);
      console.log(`  ✓ API处理成功，正在下载...`);
      
      // 下载并保存增强后的图片
      const outputPath = path.join(OUTPUT_DIR, category, file);
      await downloadImage(enhancedUrl, outputPath);
      
      console.log(`  ✓ 已保存到: ${outputPath}\n`);
      successCount++;
      
      // 延迟2秒，避免API调用过快
      if (i < images.length - 1) {
        console.log(`  等待2秒后继续...\n`);
        await delay(2000);
      }
      
    } catch (error) {
      console.error(`  ✗ 处理失败: ${error.message}\n`);
      failCount++;
      failedImages.push({ category, file, error: error.message });
      
      // 失败后等待更长时间
      if (i < images.length - 1) {
        console.log(`  等待5秒后继续...\n`);
        await delay(5000);
      }
    }
  }
  
  // 输出统计信息
  console.log('\n========================================');
  console.log('处理完成！');
  console.log('========================================');
  console.log(`总计: ${images.length} 张图片`);
  console.log(`成功: ${successCount} 张`);
  console.log(`失败: ${failCount} 张`);
  
  if (failedImages.length > 0) {
    console.log('\n失败的图片列表:');
    failedImages.forEach(({ category, file, error }) => {
      console.log(`  - ${category}/${file}: ${error}`);
    });
  }
  
  console.log(`\n增强后的图片保存在: ${OUTPUT_DIR}`);
}

// 运行脚本
main().catch(console.error);

File: enhance_images.py
# -*- coding: utf-8 -*-
"""
豆包seedream4.5图片增强脚本
批量处理equipment目录下的装备图片，使其更清晰
"""

import os
import base64
import requests
import json
import time
from pathlib import Path

# API配置
API_KEY = "ark-723d46c8-07f1-4084-8750-5d45cb53eae5-c67eb"
API_URL = "https://ark.cn-beijing.volces.com/api/v3/images/generations"
MODEL = "doubao-seedream-4-5-251128"

# 图片目录配置
EQUIPMENT_DIR = r"D:\life\code\moyu\public\images\equipment"
OUTPUT_DIR = r"D:\life\code\moyu\public\images\equipment_enhanced"

# 请求头
HEADERS = {
    "Content-Type": "application/json",
    "Authorization": f"Bearer {API_KEY}"
}

def image_to_base64(image_path):
    """
    将图片转换为base64格式
    
    Args:
        image_path: 图片路径
    
    Returns:
        base64编码的图片字符串
    """
    with open(image_path, "rb") as image_file:
        return base64.b64encode(image_file.read()).decode('utf-8')

def enhance_image(image_path, output_path, equipment_type, level):
    """
    调用豆包API增强图片清晰度
    
    Args:
        image_path: 输入图片路径
        output_path: 输出图片路径
        equipment_type: 装备类型
        level: 装备等级
    
    Returns:
        bool: 是否成功
    """
    try:
        # 将图片转换为base64
        image_base64 = image_to_base64(image_path)
        
        # 构建提示词，强调清晰度和细节
        prompt = f"高清{equipment_type}装备图标，等级{level}，清晰锐利，细节丰富，游戏装备图标风格，高质量，无水印"
        
        # 构建请求数据
        payload = {
            "model": MODEL,
            "prompt": prompt,
            "image": f"data:image/png;base64,{image_base64}",  # 使用base64格式
            "sequential_image_generation": "disabled",
            "response_format": "url",
            "size": "2K",
            "stream": False,
            "watermark": False
        }
        
        # 发送请求
        response = requests.post(API_URL, headers=HEADERS, json=payload, timeout=60)
        
        # 检查响应状态
        if response.status_code == 200:
            result = response.json()
            
            # 获取生成的图片URL
            if 'data' in result and len(result['data']) > 0:
                image_url = result['data'][0]['url']
                
                # 下载生成的图片
                img_response = requests.get(image_url, timeout=30)
                if img_response.status_code == 200:
                    # 确保输出目录存在
                    os.makedirs(os.path.dirname(output_path), exist_ok=True)
                    
                    # 保存图片
                    with open(output_path, 'wb') as f:
                        f.write(img_response.content)
                    
                    print(f"✓ 成功处理: {equipment_type}/lv{level}")
                    return True
                else:
                    print(f"✗ 下载图片失败: {equipment_type}/lv{level}")
                    return False
            else:
                print(f"✗ API返回数据格式错误: {equipment_type}/lv{level}")
                print(f"  响应: {result}")
                return False
        else:
            print(f"✗ API请求失败: {equipment_type}/lv{level}")
            print(f"  状态码: {response.status_code}")
            print(f"  错误信息: {response.text}")
            return False
            
    except Exception as e:
        print(f"✗ 处理出错: {equipment_type}/lv{level}")
        print(f"  错误: {str(e)}")
        return False

def process_all_images():
    """
    批量处理所有装备图片
    """
    print("=" * 60)
    print("豆包seedream4.5图片增强脚本")
    print("=" * 60)
    print(f"输入目录: {EQUIPMENT_DIR}")
    print(f"输出目录: {OUTPUT_DIR}")
    print("=" * 60)
    
    # 统计信息
    total_count = 0
    success_count = 0
    fail_count = 0
    
    # 装备类型映射（英文转中文）
    equipment_names = {
        "weapon": "武器",
        "helmet": "头盔",
        "clothes": "衣服",
        "shoes": "鞋子",
        "necklace": "项链",
        "bracelet": "手镯"
    }
    
    # 遍历所有装备类型目录
    for equipment_type in os.listdir(EQUIPMENT_DIR):
        type_dir = os.path.join(EQUIPMENT_DIR, equipment_type)
        
        # 跳过非目录文件
        if not os.path.isdir(type_dir):
            continue
        
        print(f"\n处理装备类型: {equipment_names.get(equipment_type, equipment_type)}")
        print("-" * 60)
        
        # 遍历该类型下的所有图片
        for image_file in sorted(os.listdir(type_dir)):
            if not image_file.endswith('.png'):
                continue
            
            # 提取等级信息
            level = image_file.replace('lv', '').replace('.png', '')
            
            # 构建输入输出路径
            input_path = os.path.join(type_dir, image_file)
            output_path = os.path.join(OUTPUT_DIR, equipment_type, image_file)
            
            total_count += 1
            
            # 处理图片
            if enhance_image(input_path, output_path, equipment_names.get(equipment_type, equipment_type), level):
                success_count += 1
            else:
                fail_count += 1
            
            # 添加延迟，避免请求过快
            time.sleep(1)
    
    # 打印统计信息
    print("\n" + "=" * 60)
    print("处理完成!")
    print(f"总计: {total_count} 张图片")
    print(f"成功: {success_count} 张")
    print(f"失败: {fail_count} 张")
    print("=" * 60)

if __name__ == "__main__":
    process_all_images()
